gcsfs/core.py

```diff
@@ -296,13 +329,47 @@ def _call(self, method, path, *args, **kwargs):
         meth = getattr(requests, method)
         if args:
             path = path.format(*[quote_plus(p) for p in args])
-        r = meth(self.base + path, headers=self.header, params=kwargs,
-                 json=json)
+        for retry in range(RETRY_COUNT):
+            try:
+                time.sleep(2**retry-1)
+                r = meth(self.base + path, headers=self.header, params=kwargs,
+                         json=json)
+                validate_response(r, path)
+                break
+            except RuntimeError as e:
+                warnings.warn("_call caught RuntimeError: " + e.__repr__())
+                with open('/tmp/error', 'wb') as f:
+                    pickle.dump(e, f)
+                print('error saved to /tmp/error')
+                if retry + 1 == RETRY_COUNT:
+                    warnings.warn('GCS call failure, out of retries: ' + str(e))
+                    raise e
+                html_error = _asHtmlError(e)
+                if html_error :
+                    if html_error['code'] in [500, 503, 504, '500', '503', '504']:
+                        # 5xx error, let's retry
+                        warnings.warn('Retrying after GCS call failure: ' + str(e))
+                        continue
+                    else:
+                        warnings.warn('GCS call failure, HtmlError but "' + str(html_error.get('code', '')) + '" so not retrying: ' + str(e))
+                else:
+                    warnings.warn('GCS call failure, not an HtmlError so not retrying: ' + str(e))
```

Comment:
What kind of situation causes this? Not-authorised, perhaps?

Reply:
All I know is that we can't exclude the possibility. I suspect a bug in our code or in the Google code could throw a RuntimeError, or perhaps bad user input of some kind.